### ISSUE TITLE
quarto only daemonizes `julia` and `python` kernels

### ISSUE THREAD

**Issue by allefeld** [quarto-dev/quarto-cli#9100]
### Bug description

I have developed a custom kernel for Matlab based on `ipykernel`. The kernel works well for my purposes, appears to be used by a few others, I haven't changed it in months, and I use it on a regular basis for Quarto documents.

Recently I noticed that processing of Quarto documents using the kernel is very slow, even for very simple computations, and then figured out that this is due to the kernel no longer being daemonized. Every time the document is processed, the kernel and thereby Matlab is started new.

### Steps to reproduce

__This assumes that Matlab is installed.__

- install the kernel from <https://github.com/allefeld/mkernel>

- if installed in an environment (e.g. conda), activate the environment

- create a file `test.qmd` with the contents

````qmd
---
jupyter: mkernel
---

```{matlab}
2 + 2
```
````

- run `quarto preview test.qmd`

- edit and save `test.qmd`

### Expected behavior

After the first start of `quarto preview`, the kernel should be started, but on subsequent renders triggered by editing and saving the file, Quarto should re-use the kernel.

### Actual behavior

The kernel is shut down after each render and restarted on the next.

```
$ quarto preview test.qmd

Starting mkernel kernel...Done

Executing 'test.quarto_ipynb'
  Cell 1/1: ''...Done

– snip – 

Output created: test.html

Watching files for changes
Browse at http://localhost:6681/
Opening in existing browser session.

Starting mkernel kernel...Done

Executing 'test.quarto_ipynb'
  Cell 1/1: ''...Done

– snip –
  
Output created: test.html

Watching files for changes
```

I checked the logs of my kernel, and indeed it receives a shutdown request, e.g.

```
Control received: {'header': {'msg_id': '6297427c-daf6f0379e7b84702c45a4c4_36565_0', 'msg_type': 'shutdown_request', 'username': 'ca', 'session': '6297427c-daf6f0379e7b84702c45a4c4', 'date': datetime.datetime(2024, 3, 15, 21, 40, 2, 470311, tzinfo=tzutc()), 'version': '5.3'}, 'msg_id': '6297427c-daf6f0379e7b84702c45a4c4_36565_0', 'msg_type': 'shutdown_request', 'parent_header': {}, 'metadata': {}, 'content': {'restart': False}, 'buffers': []}
```

To check whether this is a general problem with Jupyter kernels, in `test.qmd` I replaced `mkernel` with `python3` and `matlab` with `python`, and the Python kernel is daemonized as expected.

To see what the difference between handling of the `python3` kernel and my `mkernel` kernel is, I used the option `--execute-debug`.

<details>
<summary>Full output of <code>quarto preview test.qmd --execute-debug</code>, including modifying <code>test.qmd</code> once.</summary>

```
- Connecting to kernel

Starting mkernel kernel...2024-03-15 22:25:05,381 - root - Level 25 - starting notebook server subprocess
2024-03-15 22:25:05,381 - root - Level 25 - starting notebook server subprocess
2024-03-15 22:25:05,642 - root - Level 25 - running notebook server subprocess
2024-03-15 22:25:05,642 - root - Level 25 - creating notebook server (tcp: /run/user/1000/jt/3c0b0e66a9a6f7a52edd)
2024-03-15 22:25:05,643 - root - Level 25 - notebook server bound to port 36549
- Kernel connection successful
2024-03-15 22:25:05,723 - root - Level 25 - handling server request
- Sending execute command to kernel
- Execute command sent, reading response
2024-03-15 22:25:05,725 - root - Level 25 - executing notebook
2024-03-15 22:25:05,725 - root - Level 25 - inside notebook_execute
2024-03-15 22:25:05,760 - root - Level 25 - read notebook
2024-03-15 22:25:05,761 - root - Level 25 - not using cache
2024-03-15 22:25:05,761 - root - Level 25 - Will attempt to create notebook
2024-03-15 22:25:05,761 - root - Level 25 - type of notebook: <class 'nbformat.notebooknode.NotebookNode'>
2024-03-15 22:25:05,761 - root - Level 25 - Creating NotebookClient
2024-03-15 22:25:14,554 - root - Level 25 - NotebookClient created
2024-03-15 22:25:14,555 - root - Level 25 - Executing cell 0
2024-03-15 22:25:14,555 - root - Level 25 - Executed cell 0
2024-03-15 22:25:14,555 - root - Level 25 - Executing cell 1
Done

Executing 'test.quarto_ipynb'
  Cell 1/1: ''...2024-03-15 22:25:21,150 - root - Level 25 - Executed cell 1
2024-03-15 22:25:21,150 - root - Level 25 - Done
2024-03-15 22:25:21,150 - root - Level 25 - Notebook execution complete
2024-03-15 22:25:21,152 - root - Level 25 - notebook not persistable (exiting server)
2024-03-15 22:25:21,152 - root - Level 25 - cleaning up server resources
2024-03-15 22:25:21,152 - root - Level 25 - exiting server
Done

- Server request complete


pandoc 
  to: html
  output-file: test.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  jupyter: mkernel
  
Output created: test.html

Watching files for changes
Browse at http://localhost:6310/
Opening in existing browser session.
- Connecting to kernel

Starting mkernel kernel...2024-03-15 22:25:48,426 - root - Level 25 - starting notebook server subprocess
2024-03-15 22:25:48,426 - root - Level 25 - starting notebook server subprocess
2024-03-15 22:25:48,722 - root - Level 25 - running notebook server subprocess
2024-03-15 22:25:48,722 - root - Level 25 - creating notebook server (tcp: /run/user/1000/jt/3c0b0e66a9a6f7a52edd)
2024-03-15 22:25:48,722 - root - Level 25 - notebook server bound to port 46199
- Kernel connection successful
- Sending execute command to kernel
2024-03-15 22:25:48,771 - root - Level 25 - handling server request
- Execute command sent, reading response
2024-03-15 22:25:48,772 - root - Level 25 - executing notebook
2024-03-15 22:25:48,772 - root - Level 25 - inside notebook_execute
2024-03-15 22:25:48,815 - root - Level 25 - read notebook
2024-03-15 22:25:48,816 - root - Level 25 - not using cache
2024-03-15 22:25:48,816 - root - Level 25 - Will attempt to create notebook
2024-03-15 22:25:48,816 - root - Level 25 - type of notebook: <class 'nbformat.notebooknode.NotebookNode'>
2024-03-15 22:25:48,816 - root - Level 25 - Creating NotebookClient
2024-03-15 22:25:57,862 - root - Level 25 - NotebookClient created
Done
2024-03-15 22:25:57,863 - root - Level 25 - Executing cell 0
2024-03-15 22:25:57,864 - root - Level 25 - Executed cell 0
2024-03-15 22:25:57,864 - root - Level 25 - Executing cell 1

Executing 'test.quarto_ipynb'
  Cell 1/1: ''...2024-03-15 22:26:04,547 - root - Level 25 - Executed cell 1
2024-03-15 22:26:04,547 - root - Level 25 - Done
2024-03-15 22:26:04,547 - root - Level 25 - Notebook execution complete
Done
2024-03-15 22:26:04,548 - root - Level 25 - notebook not persistable (exiting server)

2024-03-15 22:26:04,548 - root - Level 25 - cleaning up server resources
- Server request complete


pandoc 
  to: html
  output-file: test.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  jupyter: mkernel
  
Output created: test.html

Watching files for changes
```

</details>

The difference is that with `mkernel` compared to `python3`, there are additional lines between `Notebook execution complete` and `- Server request complete`:

```
2024-03-15 22:25:21,152 - root - Level 25 - notebook not persistable (exiting server)
2024-03-15 22:25:21,152 - root - Level 25 - cleaning up server resources
2024-03-15 22:25:21,152 - root - Level 25 - exiting server
Done
```

The important bit seems to be `notebook not persistable (exiting server)`, but unfortunately a web search turned up nothing.

### Your environment

- Debian 12.5

### Quarto check output

```
Quarto 1.5.25
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.25
      Path: /opt/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: Installation From Path
      Path: /usr/bin
      Version: 2022

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.10.12 (Conda)
      Path: /home/ca/Store/micromamba/envs/std/bin/python
      Jupyter: 5.3.1
      Kernels: python3, mkernel

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.3.3
      Path: /usr/lib/R
      LibPaths:
        - /home/ca/Store/R/4.3
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      knitr: 1.45
      rmarkdown: 2.25

[✓] Checking Knitr engine render......OK
```

**Comment by allefeld** [comment]
Since reproduction as described above needs Matlab, I checked whether the same occurs with another ipykernel-based kernel, and it does. New reproduction steps:

- install `echo_kernel` from <https://github.com/jupyter/echo_kernel>

  In my experience, the installation instructions are incomplete. I installed via `pip` into a conda environment, but the kernel was not recognized – I had to additionally run `install.py` in the package installation directory, `python3 <...>/site-packages/echo_kernel/install.py`

- if installed in an environment (e.g. conda), activate the environment

- create a file `test.qmd` with the contents

````qmd
---
jupyter: echo
---

```{echo}
2 + 2
```
````

- run `quarto preview test.qmd`

- edit and save `test.qmd`

The kernel is started new on every modification of the file. With `--execute-debug`, the message `notebook not persistable (exiting server)` appears just like for my kernel.

**Comment by cscheid** [comment]
Quarto used to be too aggressive in keeping daemons running in the past and that caused problems with users expecting different jupyter versions, etc. Quarto is now more conservative about it, and that is probably causing your ipykernel to not be persisted.

The logic behind daemonization and persistent is in https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/jupyter/notebook.py#L45-L302 specifically around the field `kernel_deps` and the exception `RestartKernel`.

**Comment by allefeld** [comment]
Thanks! Though I'm not sure what should have changed between executions. As I said, it's just reprocessing of the same Quarto document, the kernel implementation and kernel specification have not changed at all.

I am also not sure about the explanation. So before executing the notebook, the kernel is restarted if `kernel_deps` has changed. But in the log, the message `notebook not persistable (exiting server)` comes after the notebook has been executed. I therefore suspect it has more to do with the return value `persist` of `notebook_execute`.

This is determined as `hasattr(notebook_execute, "kernel_deps")` (l. 301), which in turn apparently depends on the value of `nb_kernel_dependencies` (l. 290-294), which returns something if there is at least one cell output with `output.name == 'stdout' and output.output_type == 'stream'`. I don't understand what cell outputs have to do with persistability? The comment at l. 288-289 says "picks up imports that occurred witihn the notebook cells", which seems Python-specific – my kernel is based on Python, but doesn't execute Python, and `echo_kernel` doesn't execute anything.

In summary, I don't see how I can translate what the code tells me into a concrete action. Is there something I should change either in the kernel or in using Quarto to work around this?

**Comment by cscheid** [comment]
> In summary, I don't see how I can translate what the code tells me into a concrete action. 

I don't know it either. I mentioned the file because we can't predict ahead of time how arbitrary third-party kernels are going to behave, and we almost exclusively test against the python and julia kernels.

> I don't understand what cell outputs have to do with persistability? The comment at l. 288-289 says "picks up imports that occurred witihn the notebook cells", which seems Python-specific – my kernel is based on Python, but doesn't execute Python, and echo_kernel doesn't execute anything.

The kernel is persisted when there's a truthy value of `nb_kernel_depenencies` (sic); the output check happens only for the "cleanup" cell specifically. If there's no cleanup cell then we don't execute the block inside `if cleanup_cell:` which is interpreted as falsey.

(Yes, this should be properly documented.)

> Is there something I should change either in the kernel or in using Quarto to work around this?

Your kernel could emit a cell that's recognized by `nb_cleanup_cell` so that cell is executed and produces output.

**Comment by cscheid** [comment]
I would like non-`julia`, non-`python` kernels to have access to the setup and cleanup cells that we provide to `julia` and `python` without Quarto having to interfere. I think that's the correct feature for us to implement, and then your kernel can do anything it wants to, and then it could also respond to the Quarto environment as we hard-code for `python` and `julia` kernels.

**Comment by allefeld** [comment]
I looked at the code a bit more and inserted some hacky debug messages. As expected, with my kernel, the function attribute `kernel_deps` is never set, so `persist` is always `False`.

But, I'm not sure the logic here is correct. It makes sense to restart the kernel if the kernel dependencies have changed. But I don't think it is correct behavior to not persist if there are no kernel dependencies to begin with. That just means possible changes can't be detected, it doesn't mean changes have definitely occurred. In this case it would be better to keep persistence and let the user specify `daemon: false` if they want.

For the moment, I resorted to a hacky solution: Change l. 301 to `persistent = True`. As far as I can tell, that shouldn't lead to problems because the restart logic is still present.

I don't think I can contribute directly to the necessary changes in Quarto, but if there's something I can do to debug a possible solution, let me know.

**Comment by cscheid** [comment]
Yeah, I really don't think there's anything you _can_ do at the moment, which is why I changed the issue title. I'm hopeful we can make it into something that custom kernels can declare in their kernelspecs.

**Comment by cscheid** [comment]
On further thought, you could attempt to use [notebook caching](https://quarto.org/docs/computations/caching.html#jupyter-cache); if caching is enabled and succeeds, the kernel will automatically be persisted: https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/jupyter/notebook.py#L154

**Comment by allefeld** [comment]
Great, thanks! Can you let me know with which version the feature will be shipped?

**Comment by cscheid** [comment]
It'll definitely ship in 1.5. I expect to merge today or tomorrow, we just need to decide where the documentation should go.

**Comment by allefeld** [comment]
I'm using the pre-releases, so if it's included e.g. in 1.5.26, I can try it.

**Comment by cscheid** [comment]
- release: https://github.com/quarto-dev/quarto-cli/releases/tag/v1.5.26
- docs: https://quarto.org/docs/advanced/jupyter/kernel-execution.html

**Comment by allefeld** [comment]
Thanks! I've now started implementing this for my `mkernel` for Matlab, using Quarto 1.5.27.

I'm confused about the Python setup code linked from the documentation, `setup.py`. It doesn't seem to be pure Python code, but more a template for it, because there seem to be placeholders like `{fig_width}`. On the other hand, the documentation says that document options including `fig_width` are communicated via Jupyter comms, not via the setup code. And the setup code contains placeholders which are not in the list of document options: `interactivity`, `is_shiny`, `is_dashboard`, `plotly_connected`, and `run_path`.

In my code for the Matlab kernel, can I / should I use such placeholders, or the Jupyter comms, or both?

**Comment by cscheid** [comment]
> I'm confused about the Python setup code linked from the documentation, setup.py. It doesn't seem to be pure Python code, but more a template for it, because there seem to be placeholders like {fig_width}.

That's because we have special-case treatment for the `python` and `julia` languages in the jupyter engine.

> On the other hand, the documentation says that document options including fig_width are communicated via Jupyter comms, not via the setup code. 

Your kernel should use the comms path; the setup code is specific to `julia` and `python`. It's documented there for comprehensiveness, but I don't think the docs state that your setup code will get access to those entries.

> And the setup code contains placeholders which are not in the list of document options: interactivity, is_shiny, is_dashboard, plotly_connected, and run_path.

That's because we don't want to promise support for those; they're specific to either the `python` or `julia` kernels. The documentation should be your guide for implementing this feature.

**Comment by allefeld** [comment]
Alright, thanks!

I think this should be made more explicit in the documentation though. You are right that you're not promising anything, but the way the Python and Julia code are linked to right at the start suggests that something similar will work for other kernels.

It also means that there is no need for these "cells" to be executable in the target language at all. The kernel has to detect their presence anyway and then do something, and that something could be not executing anything, or executing code which is defined elsewhere.

**Comment by allefeld** [comment]
More observations:

– It doesn't seem to be allowed to create output from the setup and cleanup cells. I.e. if the kernel sends a message of type `stream` with `"name": "stdout"` in response to the setup or cleanup cell code, Quarto says

```
ERROR:

Expecting value: line 1 column 1 (char 0)
```

Creating output is not something that should be done in these cells, but if it happens accidentally the output should just be ignored, or at least the error message made more meaningful.

Error messages (`"name": "stderr"`) do not trigger the Quarto error.

– Just the presence of the `quarto_setup_cell` file, without having implemented the `execute_result` response with metadata, already enables daemonization. Assuming this is a bug which will be fixed, I implemented the `execute_result` response with metadata anyway.

**Comment by allefeld** [comment]
This goes beyond the original topic of daemonization, but since it's related to the general documentation topic "Jupyter Kernel Execution", I'll include it here.

Quarto document options are communicated once per processing, immediately before the setup cell is executed.

– The provided dictionary `options` contains many more entries than listed as "supported options" in the documentation, including further sub-dictionaries. Does "not being supported" mean that this additional information may disappear at some point?

– Without having set it in my Quarto document, `fig_format` has the value `retina`. The Quarto documentation doesn't really explain that value, and the only thing I could find is that apparently Python's Matplotlib has that as an option, where it means 2×2 antialiasing. That's not explained in the Matplotlib documentation though, just referenced on a number of websites, e.g. [here](https://stackoverflow.com/questions/54312924/what-is-matplotlibs-retina-display-mode). In any case, it doesn't seem to be a reasonable default for arbitrary `ipykernel`-based kernels. Is that a bug which will be fixed? If not, should I treat it as equivalent to `png`?

**Comment by cscheid** [comment]
> – Without having set it in my Quarto document, `fig_format` has the value `retina`. The Quarto documentation doesn't really explain that value, and the only thing I could find is that apparently Python's Matplotlib has that as an option, where it means 2×2 antialiasing. That's not explained in the Matplotlib documentation though, just referenced on a number of websites, e.g. [here](https://stackoverflow.com/questions/54312924/what-is-matplotlibs-retina-display-mode). 

> In any case, it doesn't seem to be a reasonable default for arbitrary `ipykernel`-based kernels. Is that a bug which will be fixed? 

I mean, you're probably the very first person in the world to be wiring up a Jupyter kernel for Quarto support. Whether or not it's a reasonable default would depend on all the other kernels. Right now, I think it makes sense because that value is used meaningfully by both Jupyter and Python setup cells.

> If not, should I treat it as equivalent to `png`?

I think you should treat it however it makes sense for your kernel. Julia takes `retina` to mean `svg` (lines 19--21 of setup.jl), and Python takes it to mean however matplotlib means it. This kind of cross-language configuration will always involve a certain amount of judgment calls - we take it that your kernel will do it as well as you're willing to take it. (Previously, there was no way to see any of it, after all!)

**Comment by allefeld** [comment]
Good, I'll have `retina` mean `svg` for my kernel, too. I also noticed that if the output format is `pdf`, the default `fig_format` is `pdf`, too, which is convenient.

I'm afraid there's more…

**Comment by allefeld** [comment]
If `allow_errors: false` and an error occurs, what should the kernel do to signal the error to Quarto? Just exit?

I've had a similar problem before when trying to decide what to do if the code contains a command to exit the interpreter (`quit` or `exit` in Matlab). I finally decided to just restart the Matlab engine.

In my own documents (R and Python) I always set `allow_errors: true`, because I find looking at the error in the terminal awkward; I'd rather have it as stderr in the Quarto output.

**Comment by allefeld** [comment]
Is the kernel supposed to do something special if `cache: true`? I thought caching is handled on the level of Jupyter, and the kernel doesn't need to know about it. If caching is enabled on the document level, is the kernel even sent anything?

**Comment by allefeld** [comment]
To answer my last question, I tried it. I ran into an error, and to exclude a problem with my kernel, I tried with `echo_kernel`  – the original, not your fork, because the original should still work?

Minimal document:

````qmd
---
jupyter: echo
execute:
  cache: true
---

```{echo}
2 + 2
```
````

The first processing works without a problem:

````
$ quarto preview cache.qmd --no-browser --no-watch-inputs

Starting echo kernel...Done

Executing 'cache.quarto_ipynb'

pandoc 
  to: html
  output-file: cache.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  jupyter: echo
  
Output created: cache.html

Watching files for changes
Browse at http://localhost:5271/
````

But on reprocessing (triggered by saving the file, but without changes ), this happens:

````
Starting echo kernel...ERROR: 

expected str, bytes or os.PathLike object, not builtin_function_or_method
````

To get more information:

````
$ quarto preview cache.qmd --no-browser --no-watch-inputs --execute-debug
- Connecting to kernel

Starting echo kernel...2024-03-30 00:43:19,923 - root - Level 25 - starting notebook server subprocess
2024-03-30 00:43:19,923 - root - Level 25 - starting notebook server subprocess
2024-03-30 00:43:20,175 - root - Level 25 - running notebook server subprocess
2024-03-30 00:43:20,175 - root - Level 25 - creating notebook server (tcp: /run/user/1000/jt/9a71d08a82b73a3b4e56)
2024-03-30 00:43:20,176 - root - Level 25 - notebook server bound to port 42867
- Kernel connection successful
2024-03-30 00:43:20,274 - root - Level 25 - handling server request
- Sending execute command to kernel
- Execute command sent, reading response
2024-03-30 00:43:20,285 - root - Level 25 - executing notebook
2024-03-30 00:43:20,285 - root - Level 25 - inside notebook_execute
2024-03-30 00:43:20,321 - root - Level 25 - notebook was read
2024-03-30 00:43:20,322 - root - Level 25 - {
  "name": "echo",
  "language": "text",
  "display_name": "Echo",
  "path": "/home/ca/.local/share/jupyter/kernels/echo"
}
2024-03-30 00:43:20,322 - root - Level 25 - No text directory found in /opt/quarto/share/jupyter/lang/text
2024-03-30 00:43:20,322 - root - Level 25 - Will look for explicit quarto setup cell information in kernelspec dir
2024-03-30 00:43:20,322 - root - Level 25 - No quarto_setup_cell file found in /home/ca/.local/share/jupyter/kernels/echo
2024-03-30 00:43:20,322 - root - Level 25 - /home/ca/.local/share/jupyter/kernels/echo/quarto_setup_cell
2024-03-30 00:43:20,323 - root - Level 25 - using cache
2024-03-30 00:43:20,323 - root - Level 25 - getting cache
- Error response received
ERROR: 

expected str, bytes or os.PathLike object, not builtin_function_or_method
````

I'd be happy to create an issue for this if it is helpful.

**Comment by cscheid** [comment]
> If `allow_errors: false` and an error occurs, what should the kernel do to signal the error to Quarto? Just exit?

I think throwing an exception is how ipykernel does it, but I'm not sure, actually.

> Is the kernel supposed to do something special if `cache: true`? I thought caching is handled on the level of Jupyter, and the kernel doesn't need to know about it. 

That's correct, we use [Jupyter cache](https://jupyter-cache.readthedocs.io/en/latest/).

> If caching is enabled on the document level, is the kernel even sent anything?

I don't think so, but I'd have to test to make sure.

**Comment by cscheid** [comment]
That certainly looks like a problem somewhere, likely on our side. Please file a bug, and thank you!

**Comment by allefeld** [comment]
I now tried parameters, and to my surprise, it worked, without anything special implemented on my side. I just had to install `papermill`.

Quarto recognized the chunk options with Matlab's comment character (`%`) in my initialization chunk:

````
```{matlab}
%| tags: [parameters]

alpha = 0.1
ratio = 0.1
```
````

And running Quarto with `-P alpha:0.2 -P ratio:0.3` inserted after it a chunk which uses both Matlab's comment character and its end-of-statement `;` to suppress output.

````
``` {.matlab .cell-code}
% Injected Parameters
alpha = 0.2;
ratio = 0.3;

```
````

**Comment by allefeld** [comment]
I hope this is my last comment, sorry for abusing this issue for comments after it was closed.

__Summary of feedback on the implementation:__

-   I'm not sure about the "setup cell" and "cleanup cell" mechanism. For Python and Julia they are useful because placeholders allow specific initialization of the kernel / language interpreter, but for other Jupyter kernels there isn't really anything to do here; all the initialization happens in response to the Jupyter comms.

    I ended up just putting a long randomly generated hex string here, which is recognized in `do_execute` to send the required response including metadata, but which is never executed.

-   Currently daemonization is enabled just by the presence of the cleanup cell file in the kernel spec directory, the response from `do_execute` is not necessary. That's not really a problem, but contradicts the documentation.

__Summary of feedback on the documentation:__

-   In the "Setup and cleanup cells" section, I still think linking to the Python and Julia setup cell code is likely to create misunderstanding, at least it did for me. I recommend stressing directly after that the placeholders used in them are not supported for other Jupyter kernels.

-   The "Quarto document options" section is confusing, because it mixes (a) document options which are supported but do not require active cooperation from the kernel (`params`, `cache`), (b) document options which are supported and do require active cooperation from the kernel (`allow_errors`, `fig_*`), and (c) document options which have nothing to with kernel execution (`format`). I recommend omitting (c) and making the distinction between (a) and (b) clearer.

-   I recommend to add an explanation of how to respond correctly to `allow_errors`.

Thanks for implementing this so quickly, and I'm glad to have my Matlab kernel daemonized again, and even with some additional features!

**Comment by cscheid** [comment]
I appreciate your feedback! I think you're spot-on.

We'll probably change the implementation once we have more feedback from more folks. We try hard not to break backwards-compatibility so there should be no action required on your part when we do it, but I'll let you know in case you have further comments.

**Comment by cscheid** [comment]
@cwickham Just tagging you here to keep you in the loop. When I get our docs ready for the 1.5 release, I'll want to incorporate this feedback into the docs.

**Comment by allefeld** [comment]
FYI: I just released a [new version of my Matlab kernel](https://github.com/allefeld/mkernel/releases/tag/v1.1.0) incorporating the changes for Quarto 1.5 (daemonization and document options).

**Comment by mcanouil** [comment]
@allefeld it would be more effective to announce this in the discussion board under the "show and tell" categories.

**Comment by allefeld** [comment]
Thanks for the tip!

**Comment by allefeld** [comment]
@cscheid:
> @allefeld I'm about to merge this feature.
> 
> The docs are currently in preview mode here: https://deploy-preview-1054--mystifying-jepsen-fa4396.netlify.app/docs/advanced/jupyter/kernel-execution.html
> 
> In order for your kernel to support daemonization, it will (necessarily) need some changes, but they're hopefully easy to follow. I've also created a fork of `echo_kernel` that implements these features so it should be easy to follow: https://github.com/quarto-dev/quarto-web/quarto_echo_kernel.

Unfortunately, daemonization for non-(Python/Julia)-kernels seems to have stopped working, or at least does not work reliably.

As a reminder, I opened this issue, you implemented a mechanism by which a kernel can communicate it wants to be daemonized, I implemented it in my kernel, and it worked. I haven't changed my kernel since then, there is still a setup cell file in the kernel directory and the kernel responds `daemonize: true` to its execution.

Has something been changed in Quarto which could have affected this? I'm currently on 1.5.56.

The behavior is inconsistent: If I render a full project with many `qmd` files using the kernel, Quarto says it is restarted for each file.
```
[ 4/11] tutorial-other/index.md
[ 5/11] tutorial-fmri-spm/cvcrossmanova.qmd

Starting mkernel kernel...Done

Executing 'cvcrossmanova.quarto_ipynb'
  Cell 1/2: ''...Done
  Cell 2/2: ''...Done

[ 6/11] tutorial-other/preparation.qmd

Starting mkernel kernel...Done

Executing 'preparation.quarto_ipynb'
  Cell 1/2: ''...Done
  Cell 2/2: ''...Done
```

But when I edit and preview a single file in VS Code, the kernel is not restarted.

**Comment by cscheid** [comment]
> But when I edit and preview a single file in VS Code, the kernel is not restarted.

I believe that's the desired behavior for Quarto, given that the goal is to enable fast interactive editing, and explicitly it is not to share state across executions.

**Comment by allefeld** [comment]
Okay. I thought it behaved differently in the past, but I can work with that.

**Comment by cscheid** [comment]
If you can confirm and find a reproducible way to point out how the behavior changed, then we can definitely work from that!
